pennylane/grouping/transformations.py

```diff
@@ -0,0 +1,169 @@
+    >>> qwc_rotation(pauli_dict)
+    [RY(-1.5707963267948966, wires=[0]), RX(1.5707963267948966, wires=[2])]
+
+    Args:
+        pauli_dict (dict[int, Observable]): key is the wire (int), and value is a qml.PauliX,
+            qml.PauliY, or qml.PauliZ instance.
+
+    Raises:
+        TypeError: if any values of pauli_dict are not instances of qml.PauliX, qml.PauliY, or
+            qml.PauliZ.
+
+    """
+    paulis_with_identity = (qml.Identity, qml.PauliX, qml.PauliY, qml.PauliZ)
+    if not all(isinstance(value, paulis_with_identity) for value in pauli_dict.values()):
+        raise TypeError(
+            "All values of input pauli_dict must be either Identity, PauliX, PauliY, or PauliZ instances,"
+            " instead got values: {}.".format(pauli_dict.values())
+        )
+
+    for wire in pauli_dict:
+        if isinstance(pauli_dict[wire], qml.PauliX):
+            qml.RY(-np.pi / 2, wires=wire)
+
+        elif isinstance(pauli_dict[wire], qml.PauliY):
+            qml.RX(np.pi / 2, wires=wire)
+
+
+def diagonalize_qwc_grouping(qwc_grouping):
+    """Diagonalizes a list of mutually qubit-wise commutative Pauli strings.
+
+    **Usage example:**
+
+    >>> qwc_group = [qml.PauliX(0) @ qml.PauliZ(1),
+                     qml.PauliX(0) @ qml.PauliY(3),
+                     qml.PauliZ(1) @ qml.PauliY(3)]
+    >>> diagonalize_qwc_grouping(qwc_group)
+    ([RY(-1.5707963267948966, wires=[0]), RX(1.5707963267948966, wires=[3])],
+     [Tensor(PauliZ(wires=[0]), PauliZ(wires=[1])),
+     Tensor(PauliZ(wires=[0]), PauliZ(wires=[3])),
+     Tensor(PauliZ(wires=[1]), PauliZ(wires=[3]))])
+
+    Args:
+        qwc_grouping (list[Observable]): a list of Pauli string observables that is mutually
+            qubit-wise commutative.
+
+    Returns:
+        unitary (list[Operation]): an instance of the qwc_rotation template which diagonalizes the
+            qubit-wise commuting grouping.
+        diag_terms (list[Observable]): list of Pauli string observables diagonal in the
+            computational basis.
+
+    Raises:
+        ValueError: if any 2 elements in the input QWC grouping are not qubit-wise commutative.
+
+    """
+    m_paulis = len(qwc_grouping)
+    all_wires = Wires.all_wires([pauli_word.wires for pauli_word in qwc_grouping])
+    wire_map = {i: c for c, i in enumerate(all_wires)}
+    for i in range(m_paulis):
+        for j in range(i + 1, m_paulis):
+            if not is_qwc(
+                pauli_to_binary(qwc_grouping[i], wire_map=wire_map),
+                pauli_to_binary(qwc_grouping[j], wire_map=wire_map),
+            ):
+                raise ValueError(
+                    "{} and {} are not qubit-wise commuting.".format(
+                        qwc_grouping[i], qwc_grouping[j]
+                    )
+                )
+
+    pauli_dict = {}
+    diag_terms = []
+
+    for term in qwc_grouping:
+        if term.name != ["Identity"]:
+            diag_term = Tensor(qml.Identity(wires=0))
```

Comment:
Believe that the specific wire label in `qml.Identity(wires=0)` will have to be used